### PR TITLE
feat: Add support for Data Products

### PR DESCRIPTION
**Pull request by shirshanka**
As seen in Apr Townhall, adds support for Data Products in DataHub.

Docs:
* [here](https://docs-website-7e0kpqttp-acryldata.vercel.app/docs/dataproducts)
* [and here](https://docs-website-7e0kpqttp-acryldata.vercel.app/docs/cli#dataproduct-data-product-entity)

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 1]
```diff
@@ -183,4 +185,22 @@ private static String getFilterField(final String originalField, final boolean s
         }
         return ESUtils.toKeywordField(originalField, skipKeywordSuffix);
     }
+
+    public static Filter buildFilterWithUrns(@Nonnull Set<Urn> urns, @Nullable Filter inputFilters) {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:189]: Nice - lets make sure we have unit test coverage

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/DataProductEntitiesResolver.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  static {
+    DEFAULT_ENTITIES_INPUT.setQuery(DEFAULT_QUERY);
+    DEFAULT_ENTITIES_INPUT.setStart(DEFAULT_START);
+    DEFAULT_ENTITIES_INPUT.setCount(DEFAULT_COUNT);
+  }
+
+  private final EntityClient _entityClient;
+
+  public DataProductEntitiesResolver(final EntityClient entityClient) {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/DataProductEntitiesResolver.java:49]: Do we actually use this field on the frontend? I know for domains we ended up using search mainly

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/ListDataProductAssetsResolver.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+import static com.linkedin.datahub.graphql.resolvers.ResolverUtils.bindArgument;
+import static com.linkedin.datahub.graphql.resolvers.ResolverUtils.buildFilterWithUrns;
+
+/**
+ * Resolver responsible for getting the assets belonging to a Data Product. Get the assets from the
+ * Data Product aspect, then use search to query and filter for specific assets.
+ */
+@Slf4j
+@RequiredArgsConstructor
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/ListDataProductAssetsResolver.java:42]: Interested to understand how this one is used... vs DataProductEntitiesResolver above

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/UpdateNameResolver.java [hunk 4]
```diff
@@ -53,6 +56,8 @@ public CompletableFuture<Boolean> get(DataFetchingEnvironment environment) throw
           return updateDomainName(targetUrn, input, environment.getContext());
         case Constants.CORP_GROUP_ENTITY_NAME:
           return updateGroupName(targetUrn, input, environment.getContext());
+        case Constants.DATA_PRODUCT_ENTITY_NAME:
+          return updateDataProductName(targetUrn, input, environment.getContext());
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/UpdateNameResolver.java:60]: Nice!!

File: datahub-web-react/src/app/entity/dataProduct/DataProductEntitiesTab.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { useEntityData } from '../shared/EntityContext';
+import { EmbeddedListSearchSection } from '../shared/components/styled/search/EmbeddedListSearchSection';
+import generateUseListDataProductAssets from './generateUseListDataProductAssets';
+
+export function DataProductEntitiesTab() {
+    const { urn } = useEntityData();
+
+    return (
+        <EmbeddedListSearchSection
+            useGetSearchResults={generateUseListDataProductAssets({ urn })}
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/dataProduct/DataProductEntitiesTab.tsx:11]: Hmm.... Idk of this works because listDataProducts may not accept other filter types... Can we just reuse search with a fixedFilter? Or is that not possible?

File: datahub-web-react/src/app/entity/domain/DataProductsTab/EditDataProductModal.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import { DataProduct } from '../../../../types.generated';
+import { MODAL_BODY_STYLE, MODAL_WIDTH } from './CreateDataProductModal';
+
+type Props = {
+    dataProduct: DataProduct;
+    onClose: () => void;
+    onUpdateDataProduct: (dataProduct: DataProduct) => void;
+};
+
+export default function EditDataProductModal({ dataProduct, onUpdateDataProduct, onClose }: Props) {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/domain/DataProductsTab/EditDataProductModal.tsx:15]: Do we need separate modal for create and edit?

File: datahub-web-react/src/app/entity/shared/EntityDropdown/utils.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export function isDeleteDisabled(entityType: EntityType, entityData: GenericEntityProperties | null) {
+    if (entityType === EntityType.GlossaryTerm || entityType === EntityType.GlossaryNode) {
+        const entityHasChildren = !!entityData?.children?.total;
+        const canManageGlossaryEntity = !!entityData?.privileges?.canManageEntity;
+        const canDeleteGlossaryEntity = !entityHasChildren && canManageGlossaryEntity;
+        return !canDeleteGlossaryEntity;
+    }
+    if (entityType === EntityType.DataProduct) {
+        return false; // TODO: update with permissions
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/EntityDropdown/utils.ts:12]: Do we want to handle in this PR or a folllowup?

File: datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx [hunk 8]
```diff
@@ -54,7 +54,8 @@ export default function ExternalUrlButton({ externalUrl, platformName, entityTyp
                 rel="noreferrer noopener"
                 onClick={sendAnalytics}
             >
-                View in {displayedName} <ArrowRightOutlined style={{ fontSize: 12 }} />
+                {displayedName ? `View in ${displayedName}` : 'View external link'}{' '}
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx:57]: Curious why we are doing this one? 

View external link doesn't feel very natural.. maybe just me

File: datahub-web-react/src/app/entity/shared/components/styled/ExpandedOwner/OwnerContent.tsx [hunk 9]
```diff
@@ -1,30 +1,43 @@
...
+                <TextWrapper fontSize={fontSize}>{name}</TextWrapper>
+            ) : (
                 <Popover
                     overlayStyle={{ maxWidth: 200 }}
                     placement="left"
                     title={<Typography.Text strong>{getNameFromType(owner.type)}</Typography.Text>}
                     content={<Typography.Text type="secondary">{getDescriptionFromType(owner.type)}</Typography.Text>}
                 >
-                    {name}
+                    <TextWrapper fontSize={fontSize}>{name}</TextWrapper>
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/components/styled/ExpandedOwner/OwnerContent.tsx:38]: Why in this pR?

File: datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx [hunk 10]
```diff
@@ -123,6 +124,7 @@ export const EmbeddedListSearch = ({
     shouldRefetch,
     resetShouldRefetch,
 }: Props) => {
+    const { shouldRefetchEmbeddedListSearch, setShouldRefetchEmbeddedListSearch } = useEntityContext();
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx:127]: This scares me. We use embeddedlistserach outside of entity contexts - e.g. for showing teh result of an ingestion run

Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx:127]: Curious why we are needing this now

File: datahub-web-react/src/app/entity/shared/containers/profile/header/EntityHeader.tsx [hunk 11]
```diff
@@ -66,6 +66,8 @@ export function getCanEditName(
             return privileges?.manageGlossaries || !!entityData?.privileges?.canManageEntity;
         case EntityType.Domain:
             return privileges?.manageDomains;
+        case EntityType.DataProduct:
+            return true; // TODO: add permissions for data products
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/containers/profile/header/EntityHeader.tsx:70]: Followup?

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/DataProduct/DataProductSection.tsx [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    function removeDataProduct() {
+        batchSetDataProductMutation({ variables: { input: { resourceUrns: [urn, ...siblingUrns] } } })
+            .then(() => {
+                message.success({ content: 'Removed Data Product.', duration: 2 });
+                setDataProduct(null);
+            })
+            .catch((e: unknown) => {
+                message.destroy();
+                if (e instanceof Error) {
+                    message.error({ content: `Failed to remove data product: \n ${e.message || ''}`, duration: 3 });
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/containers/profile/sidebar/DataProduct/DataProductSection.tsx:47]: Recently ive been hiding the raw error message and just writing something like:

Failed to remove Data Product. An unknown error occurred.

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx [hunk 13]
```diff
@@ -53,31 +74,34 @@ export const SidebarDomainSection = ({ readOnly }: Props) => {
         <div>
             <div id={ENTITY_PROFILE_DOMAINS_ID} className="sidebar-domain-section">
                 <SidebarHeader title="Domain" />
-                <div>
+                <ContentWrapper displayInline={!!domain}>
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx:77]: I think I may be missing why we need these changes!

File: datahub-web-react/src/app/entity/shared/entity/EntityActions.tsx [hunk 14]
```diff
@@ -105,6 +116,41 @@ function EntityActions(props: Props) {
...
+                        refetchForEntity?.();
+                        setShouldRefetchEmbeddedListSearch?.(true);
+                    }, 3000);
+                }
+            })
+            .catch((e) => {
+                message.destroy();
+                message.error(
+                    handleBatchError(entityUrns, e, {
+                        content: `Failed to add assets to Data Product: \n ${e.message || ''}`,
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/shared/entity/EntityActions.tsx:147]: same thing with this error message

File: datahub-web-react/src/app/preview/DefaultPreviewCard.tsx [hunk 15]
```diff
@@ -329,12 +335,14 @@ export default function DefaultPreviewCard({
                         </NoMarkdownViewer>
                     </DescriptionContainer>
                 )}
-                {(domain || hasGlossaryTerms || hasTags) && (
+                {(dataProduct || domain || hasGlossaryTerms || hasTags) && (
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/preview/DefaultPreviewCard.tsx:338]: Oh boy this is wild

File: datahub-web-react/src/app/shared/deleteUtils.ts [hunk 16]
```diff
@@ -23,6 +26,12 @@ export const getEntityProfileDeleteRedirectPath = (type: EntityType) => {
         case EntityType.GlossaryTerm:
             // Return to glossary page.
             return '/glossary';
+        case EntityType.DataProduct:
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/shared/deleteUtils.ts:29]: Nice so data products are ALWAYS part of a domain

File: datahub-web-react/src/app/shared/tags/TagTermGroup.tsx [hunk 17]
```diff
@@ -103,6 +105,7 @@ export default function TagTermGroup({
                         highlightText={highlightText}
                         onOpenModal={onOpenModal}
                         refetch={refetch}
+                        fontSize={fontSize}
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/shared/tags/TagTermGroup.tsx:108]: So much font size!

File: metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+   * @param description optional description of the DataProduct
+   *
+   * @return the urn of the newly created DataProduct
+   */
+  public Urn createDataProduct(
+      @Nullable String name,
+      @Nullable String description,
+      @Nonnull Authentication authentication) {
+
+    // 1. Generate a unique id for the new DataProduct.
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java:69]: Should we allow a user to provide a unique id for their product? Looks likke shirshanka's YAML format would allow that

File: metadata-models/src/main/pegasus/com/linkedin/common/Edge.pdl [hunk 19]
```diff
@@ -1,13 +1,15 @@
 namespace com.linkedin.common
 
 /**
- * Information about a relatonship edge.
- */
+* A common structure to represent all edges to entities when used inside aspects as collections
+* This ensures that all edges have common structure around audit-stamps and will support PATCH, time-travel automatically.
+**/
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/common/Edge.pdl:6]: Do we need this?

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 20]
```diff
@@ -0,0 +1,525 @@
...
+        domain (str): The domain that the Data Product belongs to. Either as a name or a fully-qualified urn.
+        owners (Optional[List[str, Ownership]]): A list of owners and their types.
+        display_name (Optional[str]): The name of the Data Product to display in the UI
+        description (Optional[str]): A documentation string for the Data Product
+        tags (Optional[List[str]]): An array of tags (either bare ids or urns) for the Data Product
+        terms (Optional[List[str]]): An array of terms (either bare ids or urns) for the Data Product
+        assets (List[str]): An array of entity urns that are part of the Data Product
+    """
+
+    class Ownership(ConfigModel):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:109]: nested classes is an antipattern - can we make this top-level?

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 21]
```diff
@@ -0,0 +1,525 @@
...
+    terms: Optional[List[str]] = None
+    properties: Optional[Dict[str, str]] = None
+    external_url: Optional[str] = None
+
+    @pydantic.validator("assets", each_item=True)
+    def assets_must_be_urns(cls, v: str) -> str:
+        try:
+            Urn.create_from_string(v)
+        except Exception as e:
+            raise ValueError(f"asset {v} is not an urn", e)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:149]:
```suggestion
            raise ValueError(f"asset {v} is not an urn: {e}") from e
```

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 22]
```diff
@@ -0,0 +1,525 @@
...
+            return self.id
+        else:
+            return f"urn:li:dataProduct:{self.id}"
+
+    # If domain is an urn, we cache it in the private _resolved_domain_urn field
+    # Otherwise, we expect the caller to populate this by using an external DomainRegistry
+    @pydantic.validator("domain")
+    def cache_resolved_domain_urn_if_urn(cls, v: str) -> str:
+        if isinstance(v, str) and v.startswith("urn:li:domain:"):
+            cls._resolved_domain_urn = v
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:165]: domain would be an instance variable but it looks like we're caching it in a class variable here?

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 23]
```diff
@@ -0,0 +1,525 @@
...
+            emitter.emit(mcp, callback)
+
+    @classmethod
+    def from_yaml(
+        cls,
+        file: Path,
+        graph: DataHubGraph,
+    ) -> "DataProduct":
+        with open(file) as fp:
+            yaml = YAML(typ="rt")  # default, if not specfied, is 'rt' (round-trip)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:283]: here using ruamel.yaml doesn't preserve anything over just using normal yaml load

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 24]
```diff
@@ -0,0 +1,525 @@
...
+        urn: str,
+        system_metadata: Optional[SystemMetadataClass] = None,
+        audit_header: Optional[KafkaAuditHeaderClass] = None,
+    ) -> DataProductPatchBuilder:
+        return DataProductPatchBuilder(
+            urn=urn, system_metadata=system_metadata, audit_header=audit_header
+        )
+
+
+DataProduct.update_forward_refs()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:525]: why is this necessary?

**Review by jjoyce0510 — COMMENTED** [review]
Overall looks great. To align with what we've been doing on other features, would definitely be hoping to see Java unit tests (Resolvers, Services, Types)

File: metadata-ingestion/src/datahub/ingestion/graph/client.py [hunk 25]
```diff
@@ -717,6 +718,23 @@ def soft_delete_urn(
...
+        try:
+            key_aspect_class = KEY_ASPECTS.get(entity_urn_parsed.get_type())
+            if key_aspect_class:
+                result = self.get_aspect(entity_urn, key_aspect_class)
+                return result is not None
+            else:
+                raise Exception(
+                    f"Failed to find key class for entity type {entity_urn_parsed.get_type()} for urn {entity_urn}"
+                )
+        except Exception as e:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/graph/client.py:732]: this exception handling might be too broad - if we get a connection error, we really should be raising an exception and not swallowing it and returning false

File: metadata-ingestion/src/datahub/specific/ownership.py [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+from datahub.metadata.schema_classes import (
+    OwnerClass,
+    OwnershipClass,
+    OwnershipTypeClass,
+)
+
+T = TypeVar("T", bound=MetadataPatchProposal)
+
+
+class OwnershipPatchHelper(Generic[T]):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/specific/ownership.py:13]: I think this one probably should be a mixin class or something - the composition interface is a bit weird

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py [hunk 27]
```diff
@@ -0,0 +1,525 @@
...
+        original_ownership_list: Optional[List[Any]],
+    ) -> Tuple[bool, Optional[List[Any]]]:
+        new_owner_type_map = {}
+        for new_owner in self.owners or []:
+            if isinstance(new_owner, DataProduct.Ownership):
+                new_owner_type_map[new_owner.id] = new_owner.type
+            else:
+                new_owner_type_map[new_owner] = "TECHNICAL_OWNER"
+        owners_matched = set()
+        patches: Dict[str, Union[dict, list]] = {"ADD": [], "DROP": {}, "REPLACE": {}}
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:354]: what if we broke this into 3 different variables - I think the resulting code would be much more readable and wouldn't need all of the type ignores

File: metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py (hunk 23 above)
Comment (shirshanka) [reply metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py:283]: made a change to preserve this original dictionary, so that it is used on the way back out in case of a diff.

File: metadata-ingestion/src/datahub/specific/ownership.py (hunk 26 above)
Comment (shirshanka) [reply metadata-ingestion/src/datahub/specific/ownership.py:13]: leaving this in for now, can get to this later.

File: metadata-ingestion/tests/test_helpers/graph_helpers.py [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+
+    def import_file(self, file: Path) -> None:
+        file_source: GenericFileSource = GenericFileSource(
+            ctx=PipelineContext(run_id="test"), config=FileSourceConfig(path=str(file))
+        )
+        for wu in file_source.get_workunits():
+            if isinstance(wu, MetadataWorkUnit):
+                metadata = wu.get_metadata().get("metadata")
+                if isinstance(
+                    metadata, (MetadataChangeProposal, MetadataChangeProposalWrapper)
```
[the remark this answers is not in the document]
Comment (shirshanka) [reply metadata-ingestion/tests/test_helpers/graph_helpers.py:52]: made it work for mce-s too

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/DataProductEntitiesResolver.java (hunk 2 above)
Comment (chriscollins3456) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/DataProductEntitiesResolver.java:49]: oh woops this is a remnant before some changes happened! will remove

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/ListDataProductAssetsResolver.java (hunk 3 above)
Comment (chriscollins3456) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataproduct/ListDataProductAssetsResolver.java:42]: yupp this one is actually used in the entities tab of a data product while data products entities resolver isn't used anymore :)

File: datahub-web-react/src/app/entity/dataProduct/DataProductEntitiesTab.tsx (hunk 5 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/dataProduct/DataProductEntitiesTab.tsx:11]: listDataProducts does accept other filter types! it takes in `SearchAcrossEntitiesInput` for queries and filters on the list of assets

File: datahub-web-react/src/app/entity/domain/DataProductsTab/EditDataProductModal.tsx (hunk 6 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/domain/DataProductsTab/EditDataProductModal.tsx:15]: hm probably not. right now they contain their own mutations separately and reuse the same `DataProductBuilderForm` component so there isn't much duplication. Can refactor if desired otherwise I think it's alright (could do in a followup as well)

File: datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx (hunk 10 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx:127]: yeah that's true but I make sure that I do a check for `if (shouldRefetchEmbeddedListSearch)` before doing anything and always assume these could be undefined so other places outside of the entity context shouldn't be affected.

the reason it was needed now is that when you click "add to assets" on a data product profile page (or domain or glossary term) then you don't see assets added in their entities tab until after a refresh of the whole page. This will update that list automatically for you for all of them.

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx (hunk 13 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx:77]: yeah the main reason for these changes are that we don't want people to be able to delete a domain off of their data product, but we do want to allow them to change it if they want. So data product profile pages will have the domain name and the "set domain" button next to it. This is just CSS updates to make it look nicer

File: datahub-web-react/src/app/shared/deleteUtils.ts (hunk 16 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/shared/deleteUtils.ts:29]: yup exactly

File: datahub-web-react/src/app/preview/DefaultPreviewCard.tsx (hunk 15 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/preview/DefaultPreviewCard.tsx:338]: yeah this component could use some TLC tbh

File: metadata-models/src/main/pegasus/com/linkedin/common/Edge.pdl (hunk 19 above)
Comment (chriscollins3456) [reply metadata-models/src/main/pegasus/com/linkedin/common/Edge.pdl:6]: need this doc string change?

File: metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java (hunk 18 above)
Comment (chriscollins3456) [reply metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java:69]: ahh good point - I could definitely see it, but maybe come back to this as a followup since i don't think that's essential in V1

File: datahub-web-react/src/app/entity/shared/containers/profile/header/EntityHeader.tsx (hunk 11 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/containers/profile/header/EntityHeader.tsx:70]: yeah same here I think this is a good followup piece with the one above since we do handle the check on the backend and throw an understandable error if they try to edit without being able to do so

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/UpdateNameResolver.java [hunk 29]
```diff
@@ -53,6 +56,8 @@
           return updateDomainName(targetUrn, input, environment.getContext());
         case Constants.CORP_GROUP_ENTITY_NAME:
           return updateGroupName(targetUrn, input, environment.getContext());
+        case Constants.DATA_PRODUCT_ENTITY_NAME:
```
[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/UpdateNameResolver.java:59]: this is incorrect.. it is reachable?

File: datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx (hunk 8 above)
Comment (shirshanka) [reply datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx]: `View source`

File: datahub-web-react/src/app/entity/shared/EntityDropdown/utils.ts (hunk 7 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/EntityDropdown/utils.ts:12]: i think followup is okay as long as we keep track!

File: datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx (hunk 8 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/ExternalUrlButton.tsx]: I'm okay with View source...

File: datahub-web-react/src/app/entity/shared/components/styled/ExpandedOwner/OwnerContent.tsx (hunk 9 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/components/styled/ExpandedOwner/OwnerContent.tsx:38]: Got it!

File: datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx (hunk 10 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearch.tsx:127]: Nice- that makes sense

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx (hunk 13 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SidebarDomainSection.tsx:77]: ahhh! okay got it

File: metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java (hunk 18 above)
Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java:69]: okay - so long as we keep track !

**Review by jjoyce0510 — APPROVED** [review]
Will ship since this PR is already super long, but do want to make note of some followups:

- Resolver Unit Tests
- DataProductType Unit Test
- DataProductService Unit Test
- Check permissions for managing data products in the UI
